src/libstore/ssh.cc

```diff
@@ -16,10 +17,21 @@ SSHMaster::SSHMaster(const std::string & host, const std::string & keyFile, bool
 
 void SSHMaster::addCommonSSHOpts(Strings & args)
 {
+    auto state(state_.lock());
+    if (!state->tmpDir)
+        state->tmpDir = std::make_unique<AutoDelete>(createTempDir("", "nix", true, true, 0700));
+
     for (auto & i : tokenizeString<Strings>(getEnv("NIX_SSHOPTS").value_or("")))
         args.push_back(i);
     if (!keyFile.empty())
         args.insert(args.end(), {"-i", keyFile});
+    if (!sshPublicHostKey.empty()) {
+        Path fileName = (Path) *state->tmpDir + "/host-key";
```

Comment:
Small thing: I think `static_cast<Path>(...)` is preferred in C++ over the C-style `(Path) ...` casts. Though for `Path` specifically, all other parts of the codebase use the C-style one already.

Also TIL that you can define castings in C++ like https://github.com/NixOS/nix/blob/199081ad00e6ee4c704eaac34211b454fe0f310c/src/libutil/util.hh#L184

Reply:
Unless I'm missing something, this cast is just a nasty way of accessing the (otherwise private) `state->tmpDir->path` field − by abusing the fact that as it's the first member of the class it has the same address as the class itself (so a `static_cast` wouldn't work because you're not supposed to be safely able to do that. And a `dynamic_cast` would fail too).

> Also TIL that you can define castings in C++ like
> ```
> operator Path() const { return path; }
> ```

Technically speaking that's not casting, it's just the definition of the `operator()` method which is the one called if you do `(*state->tmpDir)()`. But it looks indeed like it's the properly intended way of accessing the underlying path.

To bottom-line:

```suggestion
        Path fileName = (*state->tmpDir)() + "/host-key";
```

Reply:
@regnat actually there *is* a way to define conversion operators, see [here](https://en.cppreference.com/w/cpp/language/cast_operator). Those can be used with the `*_cast<T>` familly.

Reply:
I stand corrected then, looks like I was wrong all the way down.
So this is indeed a conversion operator, and although it isn't obvious from the cppreference page, `(Path)foo` actually calls `foo.operator Path()` (as shown [here](https://godbolt.org/z/qozse5)).

*(I knew I shouldn't try to pretend that I know C++ :upside_down_face: )*